controlplane/src/core/controllers/scim.ts

```diff
@@ -0,0 +1,408 @@
+import { PostgresJsDatabase } from 'drizzle-orm/postgres-js';
+import { FastifyPluginCallback, FastifyRequest } from 'fastify';
+import fp from 'fastify-plugin';
+import * as schema from '../../db/schema.js';
+import { OrganizationRepository } from '../repositories/OrganizationRepository.js';
+import { UserRepository } from '../repositories/UserRepository.js';
+import ApiKeyAuthenticator, { ApiKeyAuthContext } from '../services/ApiKeyAuthenticator.js';
+import Keycloak from '../services/Keycloak.js';
+
+export type ScimControllerOptions = {
+  db: PostgresJsDatabase<typeof schema>;
+  organizationRepository: OrganizationRepository;
+  userRepository: UserRepository;
+  authenticator: ApiKeyAuthenticator;
+  keycloakClient: Keycloak;
+  keycloakRealm: string;
+};
+
+declare module 'fastify' {
+  interface FastifyRequest {
+    authContext: ApiKeyAuthContext;
+  }
+}
+
+const plugin: FastifyPluginCallback<ScimControllerOptions> = function Scim(fastify, opts, done) {
+  fastify.addContentTypeParser('application/scim+json', { parseAs: 'string' }, function (req, body, done) {
+    try {
+      const json = JSON.parse(body.toString());
+      done(null, json);
+    } catch (err: any) {
+      done(err);
+    }
+  });
+
+  fastify.addHook('preHandler', async (req, res) => {
+    const authorization = req.headers.authorization;
+    if (!authorization) {
+      return res.code(401).send({
+        schemas: ['urn:ietf:params:scim:api:messages:2.0:Error'],
+        detail: 'Missing Authorization header.',
+        status: 401,
+      });
+    }
+    const token = authorization.replace(/^bearer\s+/i, '');
+    const authContext = await opts.authenticator.authenticate(token);
+    await opts.keycloakClient.authenticateClient();
+
+    req.authContext = authContext;
+  });
+
+  fastify.get('/', (req, res) => {
+    return res.code(200).send('SCIM');
+  });
+
+  fastify.get<{ Querystring: { filter?: string; start_index: number; count?: number } }>('/Users', async (req, res) => {
+    const authContext = req.authContext;
+
+    const filter = req.query?.filter;
+    const startIndex = req.query?.start_index;
+    const count = req.query?.count;
+
+    const users: { id: string; email: string; userName: string; active: boolean }[] = [];
+
+    if (filter) {
+      const emailFilter = filter.split(' ')[2];
```

Comment:
Check for length before accessing the index.

Reply:
Done